### ISSUE TITLE
Extensions API: enhance DecorationOptions with a custom element

### ISSUE THREAD

**Issue by wmaurer** [Microsoft/vscode#792]
vscode.DecorationOptions allows for a hoverString, but I would like decorated ranges to display a text popover text that doesn't require hovering.

This would allow me to develop an extensions equivalent to atom's [jumpy](https://atom.io/packages/jumpy)

![](https://raw.githubusercontent.com/DavidLGoldberg/jumpy/master/_images/jumpy.gif)

**Comment by wmaurer** [comment]
@alexandrudima To further refine what I'd like to be able to do:
* instead of a hover text, be able to be able to define a html string, e.g. `"<div class='jumpy label'>aa</div>"`
* be able to define the `jumpy` and `label` CSS classes
* also allow inline styles on the `div` element
* be able to position the `div` element relative to the decorated range, i.e. horizontal and vertical offset
* should also work with ranges of zero length, i.e. `Range(new Position(0, 3), new Position(0,3))`

Any feedback on the likelihood of this being implemented would be appreciated.

**Comment by spywhere** [comment]
With this such feature, I could also implement the editor rulers via extension (if the VSC is not going to implement #1673 ) properly. 

Instead of this...
![screen shot 2559-01-31 at 13 26 49](https://cloud.githubusercontent.com/assets/1087399/12700663/6e3dc660-c81e-11e5-923a-f72a2dd3f7f9.png)

Could be this...
![screen shot 2559-01-31 at 13 27 16](https://cloud.githubusercontent.com/assets/1087399/12700664/6eb210ba-c81e-11e5-92d1-8650fb1e36dc.png)

**Comment by basarat** [comment]
Migrating [alm.tools](http://alm.tools/) to monaco (:heart:)  PR https://github.com/alm-tools/alm/pull/121 

I had a working version of jumpy in alm with codemirror

![](https://raw.githubusercontent.com/alm-tools/alm-tools.github.io/master/screens/jumpy.gif)

Will try and figure out how to add this to monaco :rose:

**Comment by Brakkar** [comment]
What an amazing job you are doing, I wasn't aware of the alm suite. Can't wait it to use in on vscode. tks.

**Comment by yeluoqiuzhi** [comment]
I want to write a jumpy too, but don't have any better api to show some label on document text.

**Comment by Brakkar** [comment]
Please keep us informed here guys, if you manage to get jumpy like into vscode

**Comment by DXist** [comment]
Jumpy allows just to move to the beginning of the word. It's better to support jumps to any symbol. https://github.com/easymotion/vim-easymotion is a good example of such functionality.

**Comment by wmaurer** [comment]
Ok, so it seems that since I created this issue, the API has improved somewhat, and `DecorationOptions` supports more configuration, including a `contentIconPath`.

`DecorationOptions` could still be improved, but as far as I'm concerned this issue can be closed - I'll consider raising a new issue for further improvements.

BTW, if anyone's interested, I've created a slightly buggy but usable Jumpy extension: https://marketplace.visualstudio.com/items?itemName=wmaurer.vscode-jumpy
I'll continue to work on this extension to make it better.
